R/Seurat.R

```diff
@@ -1,205 +1,815 @@
+  if (!is.null(adata$var)) {
+    obj@assays[[assay_name]] <- SeuratObject::AddMetaData(
+      obj@assays[[assay_name]],
+      metadata = adata$var
+    )
+  }
+
+  # make sure obs and var names are set properly
+  # trackstatus: class=Seurat, feature=get_obs_names, status=done
+  # trackstatus: class=Seurat, feature=get_var_names, status=done
+  colnames(obj) <- obs_names
+  rownames(obj) <- var_names
+
+  # copy other layers
+  for (i in seq_along(layer_mapping)) {
+    from <- layer_mapping[[i]]
+    to <- names(layer_mapping)[[i]]
+    if (to == "counts" || to == "data") {
+      next
+    }
+    SeuratObject::LayerData(obj, assay = assay_name, layer = to) <- adata$layers[[from]]
+  }
+
+  # copy reductions
+  # trackstatus: class=Seurat, feature=get_obsm, status=wip
+  # trackstatus: class=Seurat, feature=get_varm, status=wip
+  if (!is.null(reduction_mapping)) {
+    if (!is.list(reduction_mapping) || (length(reduction_mapping) > 0 && is.null(names(reduction_mapping)))) {
+      stop("reduction_mapping must be a named list")
     }
+    for (i in seq_along(reduction_mapping)) {
+      reduction_name <- names(reduction_mapping)[[i]]
+      reduction <- reduction_mapping[[i]]
 
-  # translate X
-  # trackstatus: class=Seurat, feature=get_X, status=wip
-  # TODO: should x_ be passed to counts or to data?
-  # TODO: creating a seurat object when th AnnData doesn't contain X or layers
-  # probably doesn't make any sense
-  x_ <-
-    if (!is.null(obj$X)) {
-      Matrix::t(obj$X)
-    } else {
-      mat <- Matrix::sparseMatrix(
-        i = integer(0),
-        p = c(0L),
-        x = integer(0),
-        dims = c(obj$n_vars(), obj$n_obs())
+      if (
+        !is.list(reduction) || is.null(names(reduction)) ||
+          !all(names(reduction) %in% c("key", "obsm", "varm")) ||
+          !all(c("key", "obsm", "varm") %in% names(reduction))
+      ) {
+        stop("each reduction must be a list with keys 'key', 'obsm', and 'varm'")
+      }
+      dr <- .to_seurat_process_reduction(
+        adata = adata,
+        key = reduction$key,
+        obsm_embedding = reduction$obsm,
+        varm_loadings = reduction$varm,
+        assay_name = assay_name
       )
-      attr(mat, "is_X_null") <- TRUE # nolint
-      mat
+      if (!is.null(dr)) {
+        obj[[reduction_name]] <- dr
+      }
     }
-  dimnames(x_) <- list(var_names_, obs_names_)
-  x_assay <- SeuratObject::CreateAssayObject(counts = x_)
+  }
 
-  # create seurat object
-  if (ncol(var_) > 0) {
-    # don't add var metadata if the data frame does not contain any columns
-    x_assay <- SeuratObject::AddMetaData(x_assay, metadata = var_)
+  # trackstatus: class=Seurat, feature=get_obsp, status=wip
+  for (i in seq_along(graph_mapping)) {
+    graph_name <- names(graph_mapping)[[i]]
+    graph <- graph_mapping[[i]]
+    if (!is.character(graph) || length(graph) != 1) {
+      stop("each graph must be a character vector of length 1")
```

Comment:
Maybe this is clearer:

```suggestion
      stop("item ", graph_name, "in graph_mapping is not a character vector of length 1")
```